[sanitizer] Remove usage of termios ioctl constants on Linux glibc since 2.41

trcrsired opened the pull request:
glibc 2.42 made all usage of termios ioctl constants strictly internal

Therefore, we remove all usage for those removed constants.

This should only apply for Linux.

Fix #149103

Reference:
https://github.com/bminor/glibc/commit/3d3572f59059e2b19b8541ea648a6172136ec42e

@fweimer-rh @tstellar

thurstond commented:
Won't this change cause a regression (failing to intercept and sanitize those ioctls) on all pre-2.42 glibc?

Also, those ioctls are listed under SANITIZER_LINUX, not SANITIZER_GLIBC, so I suspect they're used by non-glibc as well.

thurstond commented:
> > Won't this change cause a regression (failing to intercept and sanitize those ioctls) on all pre-2.42 glibc?
> 
> Yes, but it does not matter: The `ioctl` constants were not what the kernel expects because the `struct termios` definition is different from the kernel UAPI definition.

I see, thanks for the explanation!
 
> (There is an exception for POWER, which has userspace emulation in glibc based on glibc's wrong `ioctl` constants. I don't think it makes sense to keep this around for POWER's sake.)

To clarify, does this apply to PowerPC as well, or only IBM POWER? If the former, there are still PowerPC sanitizer bots (e.g., https://lab.llvm.org/buildbot/#/builders/72) and it would be desirable to add `#if !SANITIZER_PPC` (or related constants such as SANITIZER_PPC64V2 etc.) to prevent a regression.

trcrsired commented:
Hi @thurstond @fweimer-rh .

I have seen other usages of struct termios. Do we need to remove them too?

thurstond commented:
> > > I have seen other usages of struct termios. Do we need to remove them too?
> > 
> > 
> > Please remove as little as possible (only to fix known issues, or if you can been able to test the change against other affected platforms) to avoid potential regressions.
> > For example, @fweimer-rh mentioned that this change could negatively affect POWER; it's a niche case, but let's not break it unnecessarily. There's also other libcs e.g., musl: do we know for sure that the existing termios references are not needed there?
> > > I am confused. Please show me the pesudo code of one constant on how to deal with them. Guard against glibc or remove them completely or use kernel_xxx instead?
> > 
> > 
> > Guarding against glibc would work. Alternatively, maybe something like `#ifdef TCGETS` (plus a comment about glibc).
> 
> Well the problem to use TCGETS macro is that compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.h is not supposed to introduce dependencies or it would cause macro pollution.
> 
> And I do not know whether glibc on PowerPC still defines TCGETS either, since i think it does not.
> 
> I have checked the header files an I do not find KERNEL_TCGETS to use either.

How about:
```
#if SANITIZER_LINUX
...
#if !SANITIZER_GLIBC
// move the existing TCGETS stuff here
#endif
...
#endif
```
or logically equivalent rewrite?

That would make it easy to reason that it will not break MUSL on any platform, it will not break glibc on PowerPC, or any other combination we haven't thought about.

trcrsired committed:
[sanitizer] Remove usage of termios ioctl constants for glibc

glibc 2.42 made all usage of termios ioctl constants strictly internal.

Therefore, we remove all usage for those removed constants for glibc.

We keep a pesudo definition for PowerPC to make bots happy.

[sanitizer] avoid using ioctl constants for glibc above 2.41

[compiler-rt] code formatting to make CI happy

[compiler-rt] fix a missing #endif

trcrsired commented:
What about this now?

trcrsired commented on compiler-rt/lib/sanitizer_common/sanitizer_platform.h:
```diff
@@ -482,4 +482,26 @@
+#      define SANITIZER_TCGETS 0x402c7413
+#      define SANITIZER_TCSETS 0x802c7414
+#      define SANITIZER_TCSETSF 0x802c7415
+#      define SANITIZER_TCSETSW 0x802c7416
```
but you said we need to support PPC to make bots happy

trcrsired replied:
I do not think you can simply guard the deletions with !SANITIZER_PPC since i believe these constants are removed from glibc too for PPC.

trcrsired replied:
@thurstond Hi?

trcrsired replied:
i mean the best approach is to break the ppc bots for this. or Check glibc version to see whether it is smaller than 2.42

trcrsired commented:
@thurstond I have changed the patch to only use constants below 2.41, I think this won't break PPC bots right?

thurstond commented on compiler-rt/lib/sanitizer_common/sanitizer_platform.h:
```diff
@@ -482,4 +482,17 @@
+#      endif
+#    else
+#      define SANITIZER_TERMIOS_IOCTL_CONSTANTS 1
+#    endif
```
Instead of all the SANITIZER_ and __GLIBC_PREREQ conditions, would the following work instead:
```
#    ifdef TCGETS
#        define SANITIZER_TERMIOS_IOCTL_CONSTANTS 0
#      else
#        define SANITIZER_TERMIOS_IOCTL_CONSTANTS 1
#    endif
```
?

trcrsired replied:
Including those headers would expose macro definitions to the public interface, potentially contaminating it. For any project leveraging sanitizer APIs, this could introduce conflicts or breakage in otherwise valid code.

thurstond replied:
One of your earlier proposed commits (https://github.com/llvm/llvm-project/commit/d011ca7c22f682dad4cc26916c4bbd83570ae7eb) had:
```
#    else
#      define SANITIZER_TCGETS TCGETS
#      define SANITIZER_TCSETS TCSETS
#      define SANITIZER_TCSETSF TCSETSF
#      define SANITIZER_TCSETSW TCSETSW
#    endif
```
so I don't think the `#ifdef TCGETS` approach needs to introduce any extra headers?

trcrsired replied:
That does not mean TCGETS is defined at that point. You can define the macros later on.

thurstond replied:
I think TCGETS is already defined in all the places where it matters.

I tried:
```
#ifdef TCGETS
#        define SANITIZER_TERMIOS_IOCTL_CONSTANTS 0
// Debugging only
#error "No tcgets"
#      else
#        define SANITIZER_TERMIOS_IOCTL_CONSTANTS 1
#      endif
```
and it compiled `check-sanitizer` and `check-asan` fine.

thurstond replied:
The sanitizers need the interceptors to know that the struct has been written to to avoid false positives in subsequent checks e.g.,

```
struct termios tio; // Uninitialized
ioctl(fd, TCGETS, &tio);
printf("output baud rate: %u\n", tio.c_ospeed); // MemorySanitizer would have a false positive here because it doesn't know that the ioctl initialized tio
```

(since the interceptor is in sanitizer_common, it is also used in AddressSanitizer and ThreadSanitizer. I believe omitting it may result in false positives for ThreadSanitizer as well.)

thurstond reviewed (approved):
Thanks for the fix, and thanks for bearing through my bad review ideas :-)

MaskRay commented:
> @thurstond can you make ci run now? if it passes then please merge it.
> 
> plus also need to cherry pick to update release branches for llvm21, llvm20, llvm 19

llvm 19.x no longer gets new minor releases.

For llvm 20.x, llvm 20.1.8 was supposed to be the last minor release https://discourse.llvm.org/t/llvm-20-1-8-plans/87207
However, given the build breakage with new glibc we should consider making another one...

jakeegan commented:
@MaskRay Confirmed this PR passes on ppc64le-linux.

thurstond commented:
Thank you @MaskRay for the pointers and @jakeegan for testing on powerpc64!

Does anyone have any remaining concerns for landing this patch?

trcrsired commented:
@MaskRay @thurstond Time to merge the code. Thank you:)

MaskRay commented:
> This commit will be authored by 100043421+trcrsired@users.noreply.github.com.

LLVM wants contributors to use a non-hidden email address. https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223

thurstond commented:
> hi? is anyone here to merge the patch?

The buildbots (http://google.github.io/sanitizers/show_bots.html), especially the MSan ones, are in a bad state. Out of an abundance of caution, it would be good to wait until they are green again before merging.

trcrsired commented:
> > hi? is anyone here to merge the patch?
> 
> The buildbots (http://google.github.io/sanitizers/show_bots.html), especially the ASan and MSan bots, are in a bad state. Out of an abundance of caution, it would be good to wait until they are green again before merging.

ok
